Tools/cases_generator/parser.py

```diff
@@ -221,13 +228,29 @@ def cache_effect(self) -> CacheEffect | None:
 
     @contextual
     def stack_effect(self) -> StackEffect | None:
-        # IDENTIFIER [':' IDENTIFIER]
-        # TODO: Arrays, conditions
+        # IDENTIFIER
+        #   | IDENTIFIER ':' IDENTIFIER
+        #   | IDENTIFIER '[' dimension ']'
+        # TODO: Conditions
         if tkn := self.expect(lx.IDENTIFIER):
-            type = ""
             if self.expect(lx.COLON):
-                type = self.require(lx.IDENTIFIER).text
-            return StackEffect(tkn.text, type)
+                typ = self.require(lx.IDENTIFIER)
+                return StackEffect(tkn.text, typ.text)
+            elif self.expect(lx.LBRACKET):
+                if not (dim := self.dimension()):
+                    raise self.make_syntax_error("Expected dimension")
+                self.require(lx.RBRACKET)
+                return StackEffect(tkn.text, "PyObject **", dim.text.strip())
+            else:
+                return StackEffect(tkn.text)
+
+    @contextual
+    def dimension(self) -> Dimension | None:
+        tokens: list[lx.Token] = []
+        while (tkn := self.peek()) and tkn.kind != lx.RBRACKET:
+            tokens.append(tkn)
+            self.next()
+        return Dimension(lx.to_text(tokens).strip())
```

Comment:
Nitpicks:

`dimension` can't return `None`, right?

```suggestion
                self.require(lx.RBRACKET)
                return StackEffect(tkn.text, "PyObject **", dim.text.strip())
            else:
                return StackEffect(tkn.text)

    def dimension(self) -> Dimension:
        tokens: list[lx.Token] = []
        while (tkn := self.peek()) and tkn.kind != lx.RBRACKET:
            tokens.append(tkn)
            self.next()
        return Dimension(lx.to_text(tokens).strip())
```

Also, I *think* the current definition of `dimension` allows an empty dimension (like `stuff[]`). Probably worth guarding against.

Reply:
Ah, but the correct fix to the code is for `dimension()` to check that it's got at least one token and return `None` otherwise.